### PR TITLE
Improve failing test output

### PR DESCRIPTION
**Pull request by timbussmann**
* Gave each expectation a dedicated message instead generating some generic message.
* No longer list any sent message since the value of them is highly questionable as long as we can't print the send options in a useful way
* Failed expectation now throw an `ExpectationException` instead of an `Exception`.

Open for discussion:
* ~~Should we provide dedicated output text for ExpectSend ExpectPublish and ExpectReply. Currently they all share the same text which may be confusing.~~ (done)
* Should we "dump" (list) all outgoing message (categorized by Sent, Published and Replied)?

Connects to #29

### CODE REVIEW

**Comment by timbussmann** [comment]
@mat-mcloughlin @hmemcpy please review

**Comment by timbussmann** [comment]
@Particular/nservicebus-maintainers please review

File: src/NServiceBus.Testing/ExpectationException.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+﻿namespace NServiceBus.Testing
+{
+    using System;
+
+    /// <summary>
+    /// Exception representing a failed expectation.
+    /// </summary>
+    public class ExpectationException : Exception
```
Comment (andreasohlund) [inline src/NServiceBus.Testing/ExpectationException.cs]: What's the idea with a custom exception? when would a user catch it?

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectNotSendLocal.cs [hunk 2]
```diff
@@ -5,7 +5,7 @@ namespace NServiceBus.Testing.ExpectedInvocations
 
     class ExpectNotSendLocal<TMessage> : ExpectInvocation
     {
-        private readonly Func<TMessage, bool> check;
+        readonly Func<TMessage, bool> check;
```
Comment (andreasohlund) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectNotSendLocal.cs]: Move to bottom of file (run a cleanup?)

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectDelayDeliveryWith.cs [hunk 3]
```diff
@@ -19,10 +19,13 @@ public override void Validate(TestableMessageHandlerContext context)
...
             if (!sentMessages.Any(s => check(s.Message, s.Options.GetDeliveryDelay().Value)))
             {
-                Fail(sentMessages.Select(i => i.Message).ToList());
+                Fail($"Expected a message of type {typeof(TMessage).Name} to be deferred but no message matching your constraints was deferred.");
             }
         }
 
+        // Expected <TMessage> to be delayed, but no outgoing message matched with the expected configuration.
+        // Delayed <TMessage> with 00:15:00
+        // De
```
Comment (hmemcpy) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectDelayDeliveryWith.cs]: What's this?

File: src/NServiceBus.Testing/ExpectationException.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.Testing/ExpectationException.cs]: It allows users to use `Assert.Throws<ExpectationException>` to not hide other `Exception` thrown by the code path.

File: src/NServiceBus.Testing/ExpectedInvocations/ExpectDelayDeliveryWith.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.Testing/ExpectedInvocations/ExpectDelayDeliveryWith.cs]: whops, missed to remove some notes ;)

File: src/NServiceBus.Testing/ExpectationException.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Testing/ExpectationException.cs]: :+1:

**Comment by mat-mcloughlin** [comment]
With the exception of the clean up it looks good to me

**Pull request merged**